src/store/pageview.ts

```diff
@@ -17,27 +23,51 @@ class ViewStore {
     storage.set(StorageKeys.PAGE_VIEWS, this.pageViews);
   }
 
-  _formatIntoPageView(pageFeatures: PageFeature[]) {
+  _formatIntoPageView<T>(
+    pageFeatures: PageFeature<T>[]
+  ): PageView<T> | undefined {
     const ts = timeStampInSecs();
 
-    const features = pageFeatures.reduce((acc, item) => {
-      if (!item.error) {
-        acc[item.name] = item.value;
-        return acc;
+    let keywords: string[] | null = null;
+    let topicModel: TopicModelFeature | undefined = undefined;
+    const otherFeatures: Record<string, PageFeatureValue<T>> = {};
+    for (const pageFeature of pageFeatures) {
+      if (!pageFeature.error) {
+        switch (pageFeature.name) {
+          case 'keywords':
+            keywords =
+              pageFeature.value instanceof Array ? pageFeature.value : null;
+            break;
+          case 'topicModelFeatures':
+            topicModel =
+              !(pageFeature.value instanceof Array) &&
+              pageFeature.value instanceof Object
+                ? pageFeature.value
+                : undefined;
+            break;
+          default:
+            otherFeatures[pageFeature.name] = pageFeature.value;
+        }
       }
-      return acc;
-    }, {} as Record<string, string[]>);
+    }
 
-    if (Object.keys(features).length < 1) return undefined;
-
-    return {
-      ts,
-      features,
-    };
+    if (topicModel) {
+      return {
+        ts,
+        features: { topicModel, ...otherFeatures },
+      };
+    } else if (keywords) {
+      return {
+        ts,
+        features: { keywords, ...otherFeatures },
+      };
+    } else {
+      return undefined;
+    }
```

Comment:
There seems to be lots of if or case based on this - can we solve this in a better manner?

Reply:
Typescript needs the conditionals for refinement -- I ended up using [type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#typeof-type-guards) though and object spread to make it more succinct